In [1]:
import tensorflow as tf
import tensorflow.keras as keras
import pandas as pd
import numpy as np
import scipy
import matplotlib.pyplot as plt

import seaborn as sns
import matplotlib
from matplotlib import pyplot as plt
%matplotlib inline

In [2]:
# The Python magic
import IPython
import signal

def interrupted(_interrupted=[False], _default=[None]):
    if _default[0] is None or signal.getsignal(signal.SIGINT) == _default[0]:
        _interrupted[0] = False
        def handle(signal, frame):
            if _interrupted[0] and _default[0] is not None:
                _default[0](signal, frame)
            print('Interrupt!')
            _interrupted[0] = True
        _default[0] = signal.signal(signal.SIGINT, handle)
    return _interrupted[0]

def enumerate_cycle(g):
    epoch = 0
    while True:
        for i,x in enumerate(g):
            yield (epoch,i), x
        epoch = epoch + 1

In [5]:
# Loading the dataset
pd.set_option('display.max_columns', 500)

diabetic = pd.read_csv('diabetes/diabetic_data_original.csv')
diabetic

encounter_id  patient_nbr             race  gender      age weight  \
0            2278392      8222157        Caucasian  Female   [0-10)      ?   
1             149190     55629189        Caucasian  Female  [10-20)      ?   
2              64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3             500364     82442376        Caucasian    Male  [30-40)      ?   
4              16680     42519267        Caucasian    Male  [40-50)      ?   
...              ...          ...              ...     ...      ...    ...   
101761     443847548    100162476  AfricanAmerican    Male  [70-80)      ?   
101762     443847782     74694222  AfricanAmerican  Female  [80-90)      ?   
101763     443854148     41088789        Caucasian    Male  [70-80)      ?   
101764     443857166     31693671        Caucasian  Female  [80-90)      ?   
101765     443867222    175429310        Caucasian    Male  [70-80)      ?   

        admission_type_id  discharge_disposition_id  admission_source_id  \
0                       6                        25                    1   
1                       1                         1                    7   
2                       1                         1                    7   
3                       1                         1                    7   
4                       1                         1                    7   
...                   ...                       ...                  ...   
101761                  1                         3                    7   
101762                  1                         4                    5   
101763                  1                         1                    7   
101764                  2                         3                    7   
101765                  1                         1                    7   

        time_in_hospital payer_code         medical_specialty  \
0                      1          ?  Pediatrics-Endocrinology   
1                      3          ?                         ?   
2                      2          ?                         ?   
3                      2          ?                         ?   
4                      1          ?                         ?   
...                  ...        ...                       ...   
101761                 3         MC                         ?   
101762                 5         MC                         ?   
101763                 1         MC                         ?   
101764                10         MC           Surgery-General   
101765                 6          ?                         ?   

        num_lab_procedures  num_procedures  num_medications  \
0                       41               0                1   
1                       59               0               18   
2                       11               5               13   
3                       44               1               16   
4                       51               0                8   
...                    ...             ...              ...   
101761                  51               0               16   
101762                  33               3               18   
101763                  53               0                9   
101764                  45               2               21   
101765                  13               3                3   

        number_outpatient  number_emergency  number_inpatient  diag_1  diag_2  \
0                       0                 0                 0  250.83       ?   
1                       0                 0                 0     276  250.01   
2                       2                 0                 1     648     250   
3                       0                 0                 0       8  250.43   
4                       0                 0                 0     197     157   
...                   ...               ...               ...     ...     ...   
101761                  0                 0          

In [40]:
np.unique(diabetic['patient_nbr'].value_counts().values, return_counts=True)

(array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23, 28, 40]),
 array([54745, 10434,  3328,  1421,   717,   346,   207,   111,    70,
           42,    20,    19,    14,     5,     9,     4,     3,     6,
            3,     6,     1,     2,     3,     1,     1]))

In [41]:
54745/len(diabetic)

0.5379498064186466

In [45]:
(54745+10434)/len(diabetic)

0.6404791384155808

In [4]:
patient_nbr_counts = diabetic['patient_nbr'].value_counts()
df_diabetic = diabetic[diabetic['patient_nbr'].isin(patient_nbr_counts.index[patient_nbr_counts.gt(1)])].sort_values(['patient_nbr', 'encounter_id'])
pd.set_option('display.max_rows', 25)
print(df_diabetic[['encounter_id', 'patient_nbr', 'readmitted']])

        encounter_id  patient_nbr readmitted
4267        24437208          135        <30
4780        26264286          135        >30
1164         8380170         1152        >30
5953        30180318         1152        >30
14180       55533660         1152        >30
...              ...          ...        ...
99544      414418886    188634893         NO
93050      330256946    188970179        <30
101595     441488168    188970179         NO
90933      302575430    189257846         NO
98191      397823084    189257846        >30

[47021 rows x 3 columns]


Patients that have been readmitted multiple times

In [7]:
df_diabetic = df_diabetic.set_index(df_diabetic['encounter_id'])
df_diabetic['len_next_stay'] = pd.Series()
for patient in df_diabetic['patient_nbr'].unique():
    encounter_ids = list(df_diabetic[df_diabetic['patient_nbr'] == patient]['encounter_id'])
    readmitted = df_diabetic.loc[encounter_ids, ['readmitted']]['readmitted'].value_counts()
    lengths_of_stay = df_diabetic.loc[encounter_ids, 'time_in_hospital']
    df_diabetic.loc[encounter_ids[:-1], ['len_next_stay']] = lengths_of_stay.values[1:]

df_readmitted_only = df_diabetic.copy()
df_readmitted_only = df_readmitted_only[np.isfinite(df_readmitted_only['len_next_stay'])]

In [10]:
np.unique(df_readmitted_only['patient_nbr'].value_counts().values, return_counts=True)

(array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 27, 39]),
 array([10434,  3328,  1421,   717,   346,   207,   111,    70,    42,
           20,    19,    14,     5,     9,     4,     3,     6,     3,
            6,     1,     2,     3,     1,     1]))

Removing the patients that have been not readmitted more than once or have been readmitted after NO readmission outcome

In [26]:
multiple_not_readmitted = []
for patient in df_diabetic['patient_nbr'].unique():
    encounter_ids = list(df_diabetic[df_diabetic['patient_nbr'] == patient]['encounter_id'])
    readmitted = df_diabetic.loc[encounter_ids, ['readmitted']]['readmitted'].value_counts()
    if 'NO' in readmitted and readmitted['NO'] > 1:
        multiple_not_readmitted.append(patient)

len(multiple_not_readmitted)

44

In [29]:
df_readmitted_only[df_readmitted_only['readmitted'] == 'NO']

encounter_id  patient_nbr             race  gender      age  \
encounter_id                                                                
90671322          90671322      1189206  AfricanAmerican    Male  [80-90)   
82617192          82617192      1305612  AfricanAmerican  Female  [70-80)   
90400272          90400272      1405557  AfricanAmerican  Female  [70-80)   
93566430          93566430      3195774        Caucasian    Male  [70-80)   
91783026          91783026     16697187        Caucasian  Female  [70-80)   
...                    ...          ...              ...     ...      ...   
293778884        293778884    186400166        Caucasian    Male  [80-90)   
377345516        377345516    187298114        Caucasian    Male  [80-90)   
337315430        337315430    187920311        Caucasian  Female  [60-70)   
372665912        372665912    188238020        Caucasian  Female  [80-90)   
302575430        302575430    189257846            Asian    Male  [60-70)   

             weight  admission_type_id  discharge_disposition_id  \
encounter_id                                                       
90671322          ?                  2                         3   
82617192          ?                  2                         3   
90400272          ?                  2                         1   
93566430          ?                  2                         1   
91783026          ?                  2                         1   
...             ...                ...                       ...   
293778884         ?                  1                         3   
377345516         ?                  1                         1   
337315430         ?                  3                         6   
372665912         ?                  1                         6   
302575430         ?                  1                         1   

              admission_source_id  time_in_hospital payer_code  \
encounter_id                                                     
90671322                        1                 3         MC   
82617192                        1                 6         MC   
90400272                        1                 9         MC   
93566430                        4                 9         HM   
91783026                        1                 6         HM   
...                           ...               ...        ...   
293778884                       7                 3         MC   
377345516                       7                 3         MC   
337315430                       1                11         PO   
372665912                       7                 3         MC   
302575430                       7                 4         HM   

                              medical_specialty  num_lab_procedures  \
encounter_id                                                          
90671322                       InternalMedicine                  31   
82617192                       InternalMedicine                  56   
90400272                             Nephrology                  43   
93566430                             Nephrology                  44   
91783026      PhysicalMedicineandRehabilitation                  45   
...                                         ...                 ...   
293778884                                     ?                  19   
377345516                                     ?                  50   
337315430                                     ?                  37   
372665912                                     ?                  65   
302575430                                     ?                  54   

              num_procedures  num_medications  number_outpatient  \
encounter_id                                                       
90671322                   0                9                  0   
82617192                   1               26                  0   
90400272                   4               13                  0   
9356

In [27]:
to_exclude = []
for patient in df_readmitted_only['patient_nbr'].unique():
    encounter_ids = list(df_readmitted_only[df_readmitted_only['patient_nbr'] == patient]['encounter_id'])
    readmitted = df_readmitted_only.loc[encounter_ids, ['readmitted']]['readmitted'].value_counts()
    if 'NO' in readmitted:
        to_exclude.append(patient)

to_exclude

[1189206,
 1305612,
 1405557,
 3195774,
 16697187,
 17079237,
 18627624,
 23778297,
 27956088,
 29108259,
 29324601,
 29368215,
 29412864,
 30184254,
 30325419,
 30325446,
 30526470,
 30897927,
 31355928,
 31365891,
 31652982,
 31665015,
 31665987,
 32011137,
 32236452,
 32263560,
 32311764,
 32314608,
 32337963,
 32565870,
 32673303,
 33108129,
 33230016,
 33379713,
 33454008,
 33581808,
 33845625,
 33896214,
 33946560,
 34248249,
 34301979,
 34507611,
 35047287,
 35052723,
 35510715,
 35816535,
 35967501,
 36161775,
 36216972,
 36240570,
 36851895,
 37010466,
 37255248,
 37459422,
 37592973,
 38045853,
 38056068,
 38074284,
 38075310,
 38082618,
 38147922,
 38225853,
 38264805,
 38502261,
 38515230,
 38657925,
 38666502,
 38768337,
 38774187,
 38799234,
 38812941,
 39125133,
 39433527,
 39602736,
 39626514,
 39740490,
 39783186,
 39791142,
 39857706,
 39911454,
 39923199,
 40325778,
 40373325,
 40395537,
 40442940,
 40449735,
 40475232,
 40520277,
 40534695,
 40623579,
 40668219,
 40

In [30]:
len(to_exclude)

489

(the discrepancy between the above 490 rows and `to_exclude=489` is related to that one patient is still not readmitted more than once)

In [31]:
df_filtered = df_readmitted_only.drop(df_readmitted_only[df_readmitted_only['readmitted'] == 'NO'].index, axis=0)
df_filtered

encounter_id  patient_nbr             race  gender      age  \
encounter_id                                                                
24437208          24437208          135        Caucasian  Female  [50-60)   
8380170            8380170         1152  AfricanAmerican  Female  [50-60)   
30180318          30180318         1152  AfricanAmerican  Female  [50-60)   
55533660          55533660         1152  AfricanAmerican  Female  [60-70)   
80742510          80742510         1152  AfricanAmerican  Female  [60-70)   
...                    ...          ...              ...     ...      ...   
322610720        322610720    187298114        Caucasian    Male  [80-90)   
363787034        363787034    187298114        Caucasian    Male  [80-90)   
344902604        344902604    188284883            Other  Female  [80-90)   
364338650        364338650    188634893        Caucasian    Male  [70-80)   
330256946        330256946    188970179            Asian  Female  [70-80)   

             weight  admission_type_id  discharge_disposition_id  \
encounter_id                                                       
24437208          ?                  2                         1   
8380170           ?                  1                         1   
30180318          ?                  1                         1   
55533660          ?                  1                         1   
80742510          ?                  1                         1   
...             ...                ...                       ...   
322610720         ?                  1                         1   
363787034         ?                  1                         6   
344902604         ?                  1                         1   
364338650         ?                  3                         1   
330256946         ?                  1                         6   

              admission_source_id  time_in_hospital payer_code  \
encounter_id                                                     
24437208                        1                 8          ?   
8380170                         7                 6          ?   
30180318                        7                 6          ?   
55533660                        7                10          ?   
80742510                        7                 8          ?   
...                           ...               ...        ...   
322610720                       7                 2         MC   
363787034                       7                 4         MC   
344902604                       7                 1         MC   
364338650                       1                 4         MC   
330256946                       7                 4         MC   

                medical_specialty  num_lab_procedures  num_procedures  \
encounter_id                                                            
24437208               Cardiology                  77               6   
8380170       Hematology/Oncology                  43               2   
30180318      Hematology/Oncology                  45               4   
55533660      Hematology/Oncology                  54               2   
80742510                        ?                  30               1   
...                           ...                 ...             ...   
322610720                       ?                  13               0   
363787034                Resident                  71               0   
344902604                       ?                  61               2   
364338650        InternalMedicine                  52               0   
330256946                       ?                  59               1   

              num_medications  number_outpatient  number_emergency  \
encounter_id                                                         
24437208                   33                  0                 0   
8380170                    13                  0                 0   
30180318                   15         

Number of patients:

In [34]:
len(np.unique(df_filtered['patient_nbr']))

16515

Sequence lengths:

In [32]:
np.unique(df_filtered['patient_nbr'].value_counts().values, return_counts=True)

(array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 27, 39]),
 array([10280,  3276,  1403,   698,   338,   207,   110,    65,    42,
           21,    18,    14,     5,     8,     4,     3,     6,     3,
            6,     1,     2,     3,     1,     1]))

In [39]:
10280/16515

0.6224644262791402

## Handling missing values

In [35]:
for column in list(df_diabetic.columns.values):
    df_filtered[column] = np.where(df_filtered[column] == '?', np.nan, df_filtered[column])
df_filtered

/Users/kamilestankeviciute/anaconda3/envs/datasci/lib/python3.7/site-packages/pandas/core/ops/__init__.py:1115: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


encounter_id  patient_nbr             race  gender      age  \
encounter_id                                                                
24437208        24437208.0        135.0        Caucasian  Female  [50-60)   
8380170          8380170.0       1152.0  AfricanAmerican  Female  [50-60)   
30180318        30180318.0       1152.0  AfricanAmerican  Female  [50-60)   
55533660        55533660.0       1152.0  AfricanAmerican  Female  [60-70)   
80742510        80742510.0       1152.0  AfricanAmerican  Female  [60-70)   
...                    ...          ...              ...     ...      ...   
322610720      322610720.0  187298114.0        Caucasian    Male  [80-90)   
363787034      363787034.0  187298114.0        Caucasian    Male  [80-90)   
344902604      344902604.0  188284883.0            Other  Female  [80-90)   
364338650      364338650.0  188634893.0        Caucasian    Male  [70-80)   
330256946      330256946.0  188970179.0            Asian  Female  [70-80)   

             weight  admission_type_id  discharge_disposition_id  \
encounter_id                                                       
24437208        NaN                2.0                       1.0   
8380170         NaN                1.0                       1.0   
30180318        NaN                1.0                       1.0   
55533660        NaN                1.0                       1.0   
80742510        NaN                1.0                       1.0   
...             ...                ...                       ...   
322610720       NaN                1.0                       1.0   
363787034       NaN                1.0                       6.0   
344902604       NaN                1.0                       1.0   
364338650       NaN                3.0                       1.0   
330256946       NaN                1.0                       6.0   

              admission_source_id  time_in_hospital payer_code  \
encounter_id                                                     
24437208                      1.0               8.0        NaN   
8380170                       7.0               6.0        NaN   
30180318                      7.0               6.0        NaN   
55533660                      7.0              10.0        NaN   
80742510                      7.0               8.0        NaN   
...                           ...               ...        ...   
322610720                     7.0               2.0         MC   
363787034                     7.0               4.0         MC   
344902604                     7.0               1.0         MC   
364338650                     1.0               4.0         MC   
330256946                     7.0               4.0         MC   

                medical_specialty  num_lab_procedures  num_procedures  \
encounter_id                                                            
24437208               Cardiology                77.0             6.0   
8380170       Hematology/Oncology                43.0             2.0   
30180318      Hematology/Oncology                45.0             4.0   
55533660      Hematology/Oncology                54.0             2.0   
80742510                      NaN                30.0             1.0   
...                           ...                 ...             ...   
322610720                     NaN                13.0             0.0   
363787034                Resident                71.0             0.0   
344902604                     NaN                61.0             2.0   
364338650        InternalMedicine                52.0             0.0   
330256946                     NaN                59.0             1.0   

              num_medications  number_outpatient  number_emergency  \
encounter_id                                                         
24437208                 33.0                0.0               0.0   
8380170                  13.0                0.0               0.0   
30180318                 15.0         

## Generating encounter sequences